### ISSUE TITLE
Patching external resources

### ISSUE THREAD

**Issue by plietar** [Homebrew/legacy-homebrew#31508]
TL;DR : Patching resources is currently hacky
Hi,
What is the proper way to patch a downloaded resource ?
In my [formula](https://github.com/plietar/homebrew-stm32/blob/470776cfce6b691dc6c05a6691eabf434b75f5cb/arm-none-eabi-gcc.rb) I used to do the following, inside the `install` function : 
```ruby
resource('newlib').stage do
  Patch.create(:p1, :DATA).apply
  (buildpath).install Dir['newlib', 'libgloss']
end
```

However, this was broken by #31199, as the patch is created on the fly, and is not part of the formula's spec, and thus the formula's path does not get inject by [this code] (https://github.com/Homebrew/homebrew/blob/88be10455ae5209201d488c73c3e398b6b236136/Library/Homebrew/formula.rb#L582) : 

```ruby
active_spec.patches.grep(DATAPatch).each { |p| p.path = path }
```

The build fails when `DATAPatch` tries [to open a `nil` path](https://github.com/Homebrew/homebrew/blob/88be10455ae5209201d488c73c3e398b6b236136/Library/Homebrew/patch.rb#L88)

My current workaround is to set the path manually : 
```ruby
resource('newlib').stage do
  p = Patch.create(:p1, :DATA)
  p.path = Pathname.new(__FILE__).expand_path
  p.apply

  (buildpath).install Dir['newlib', 'libgloss']
end
```
However this looks hacky, and is very dependent on the DATAPatch implementation.

Is the a way of doing this more cleanly ? Adding patch support to resources' DSL would be amazing : 

```ruby
resource 'newlib' do
  url 'ftp://sourceware.org/pub/newlib/newlib-2.1.0.tar.gz'
  sha1 '364d569771866bf55cdbd1f8c4a6fa5c9cf2ef6c'
  patch :p1, :DATA
end
```

Thanks !

**Comment by yu-ichiro** [comment]
Any updates? I exactly need this.

**Comment by mikemcquaid** [comment]
@yu-ichiro None.

**Comment by jacknagel** [comment]
I have an idea for a different approach.

Instead of adding support for patches to individual resources, we provide a way to stage resources before the normal patching machinery executes. This way, patches can be generated against both the main download and any staged resources.

```rb
def pre_patch
  buildpath.install resource("foo")
end
```

This pre_patch method would be called after the main download is staged into the build directory, but before patching.

**Comment by bcomnes** [comment]
+1 to @jacknagel's idea

**Comment by dakcarto** [comment]
+1 to @jacknagel's approach. Simple solution.

**Comment by MikeMcQuaid** [comment]
We're closing out all issues relating to features that no-one is implementing in the short term in favour of pull requests.
